src/features/validation/index.ts

```diff
@@ -0,0 +1,241 @@
+  validating: () => Promise<void>;
+  visibility: Visibility;
+  setNodeVisibility: (nodes: LayoutNode[], newVisibility: number, rowIndex?: number) => void;
+  showAllErrors: boolean;
+  setShowAllErrors: (showAllErrors: boolean) => void;
+  setAttachmentVisibility: (attachmentId: string, node: LayoutNode, newVisibility: number) => void;
+  removeRowVisibilityOnDelete: (node: LayoutNode<'Group'>, rowIndex: number) => void;
+};
+
+export type ValidationState = FormValidations & {
+  task: BaseValidation[];
+};
+
+export type FormValidations = {
+  fields: FieldValidations;
+  components: ComponentValidations;
+};
+
+export type ValidationGroup<T extends GroupedValidation> = {
+  [group: string]: T[];
+};
+
+export type FieldValidations = {
+  [field: string]: ValidationGroup<FieldValidation>;
+};
+
+export type ComponentValidations = {
+  [componentId: string]: {
+    bindingKeys: { [bindingKey: string]: ValidationGroup<ComponentValidation> };
+    component: ValidationGroup<ComponentValidation>;
+  };
+};
+
+export type BaseValidation<Severity extends ValidationSeverity = ValidationSeverity> = {
+  message: TextReference;
+  severity: Severity;
+  category: ValidationCategory;
+};
+
+export type GroupedValidation<Severity extends ValidationSeverity = ValidationSeverity> = BaseValidation<Severity> & {
+  group: string;
+};
+
+export type FieldValidation<Severity extends ValidationSeverity = ValidationSeverity> = GroupedValidation<Severity> & {
+  field: string;
+};
+
+export type ComponentValidation<Severity extends ValidationSeverity = ValidationSeverity> =
+  GroupedValidation<Severity> & {
+    componentId: string;
+    bindingKey?: string;
+    meta?: Record<string, string>;
+  };
+
+export type NodeValidation<Severity extends ValidationSeverity = ValidationSeverity> = GroupedValidation<Severity> & {
+  componentId: string;
+  pageKey: string;
+  bindingKey?: string;
+  meta?: Record<string, string>;
+};
+
+export type AttachmentChange = {
+  node: LayoutNode;
+  attachmentId: string;
+};
+
+/**
+ * Contains all of the necessary elements from the redux store to run frontend validations.
+ */
+export type ValidationDataSources = {
+  currentLanguage: string;
+  formData: object;
+  attachments: IAttachments;
+  application: IApplicationMetadata;
+  instance: IInstance | null;
+  process: IProcess | null;
+  layoutSets: ILayoutSets;
+  schema: JSONSchema7;
+  customValidation: IExpressionValidations | null;
+};
+
+export type ValidationContextGenerator = (node: LayoutNode | undefined) => ValidationDataSources;
+
+/**
+ * This format is used by the backend to send validation issues to the frontend.
+ */
+export interface BackendValidationIssue {
+  code?: string;
+  description?: string;
+  field?: string;
+  dataElementId?: string;
+  severity: BackendValidationSeverity;
+  source: ValidationIssueSources;
+  customTextKey?: string;
+  customTextParams?: ValidLangParam[]; //TODO(Validation): Probably broken for text resources currently
```

Comment:
Probably not broken anymore. :-D

Reply:
Ah, maybe create an issue for it then. Would be nice to have tests to ensure it works at a later date.

Reply:
https://github.com/Altinn/app-frontend-react/issues/1723